### ISSUE TITLE
Unimplemented Unimplemented(RandomNormalLike) ToTypedTransla

### ISSUE THREAD

**Issue by stillonearth** [sonos/tract#895]
I'm trying to make Soft-Actor Critic model from stable_baselines3 work with onnx and get following error.

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Translating node #39 "RandomNormalLike_25" Unimplemented(RandomNormalLike) ToTypedTranslator

Caused by:
    0: translating op UnimplementedOp { outputs: 1, name: "RandomNormalLike", message: "NodeProto { input: [\"onnx::RandomNormalLike_102\"], output: [\"onnx::Mul_103\"], name: \"RandomNormalLike_25\", op_type: \"RandomNormalLike\", domain: \"\", attribute: [], doc_string: \"\" }" }
```

Do you plan to support `RandomNormalLike` op?

**Comment by kali** [comment]
It's not there but it should not be too hard to add. We have a couple of random uniform tensorflow ops to start from, which use a relatively cheap generator. Maybe they should be moved to core in some form or another. IIRC, we have a separate int and float generators.

I am prudent about avoiding redundancy in core. Not sure if we should mimick what I did in TF, or collapse the int and float. The Like and regular variant feel a bit redundant in onnx spec, we probably want to have only one in core (the other emulated with the Shape op).

Converting to uniform is pretty easy (Box Muller ?) so it may be a separate op in core.

**Comment by stillonearth** [comment]
So what do you suggest? Create a PR with this operator or have it implemented in side lib?

**Comment by kali** [comment]
@stillonearth Could you give me a few pointer on how to obtain these models and how to run them ? I am implementing the four Random operators from ONNX, but I am completely in the dark for testing them (ONNX test suite does not cover the Random* operators...)

**Comment by stillonearth** [comment]
An example would be: 

https://github.com/stillonearth/bevy_quadruped_neural_control/blob/main/examples/tract.rs

onnx model would be obtained from onnx exporting of [stable baselines SAC](https://stable-baselines3.readthedocs.io/en/master/modules/sac.html) policy

**Comment by kali** [comment]
Thanks. Anyway you can upload the onnx file somewhere ?

**Comment by stillonearth** [comment]
The one from an example: https://github.com/stillonearth/bevy_quadruped_neural_control/raw/main/assets/g1-control-direction.onnx

**Comment by kali** [comment]
Would you try #921 ? As far as I can tell, it works, but random ops are tricky to test.

**Comment by stillonearth** [comment]
Seems it works, thanks.

**Comment by kali** [comment]
Closing this, then. Will appear on upcoming 0.19.

**Comment by stillonearth** [comment]
Is there any way to provide a randomness seed to a tract runtime?

**Comment by kali** [comment]
You can specify the seed at the operator level. But even with the same seed, you would not get the same results, they is no reason to believe torch and tract use the same random generator.

But I think there is a bug anyway, on my side... It looks like I'm somewhat getting confused and use a Uniform law. Lemme fix that.

```
┃┃┏ 48 Random _actor_RandomNormalLike
┃┃┃   * Random { fact: 1,12,F32, dist: Uniform { low: ,F32 0, high: ,F32 1 }, seed: None }
┃┃┃   ━━━ 1,12,F32
```

**Comment by kali** [comment]
can you have a look with the branch ? see if it's better ?

**Comment by stillonearth** [comment]
I've tried #925 , but still results are different from pytorch runtime:

Here's an pytorch-run policy: 

![img1](https://user-images.githubusercontent.com/97428129/210062110-174034fa-461d-4553-996a-7271fcee75fc.gif)

Here's same nn exported to onnx and
![img2](https://user-images.githubusercontent.com/97428129/210062170-9e4f331d-0da2-4042-83bd-924a311adf9c.gif)

Is it possible to isolate RandomNormalLike op for testing?

**Comment by kali** [comment]
Lol, the poor thing looks very distressed.

I've made 1000 runs, capture the output of the random op and plotted the distribution, it looks ok to me :/

<img width="589" alt="image" src="https://user-images.githubusercontent.com/53657/210067589-def5834c-17ec-42b8-9b98-3b2b7745c626.png">

**Comment by stillonearth** [comment]
Okay here's mean of executing this NN with same input for 100 times:

**Tract**
```
[0.22173594, -0.3220566, 0.92420906, -0.029943507, -0.27507502, 0.9767548, 0.98466545, -0.24677467, 0.972273, -0.94508785, -0.38793334, 0.98633856]
```

**PyTorch**
```
[0.20535805, -0.31723608,  0.92612128, -0.03568691, -0.27401054, 0.97724471,  0.98378817, -0.24789757, 0.97501801, -0.95094981, -0.38920372,  0.98649332]
```

Results seem on-par

**Comment by kali** [comment]
Mmmm... would it be worth trying to compare the standard deviation as well ? See if we can see a difference ?

**Comment by stillonearth** [comment]
**Tract std**

```[0.30818725, 0.07063462, 0.07817909, 0.18184128, 0.11369549, 0.028545896, 0.026648697, 0.07963544, 0.03102807, 0.06768072, 0.08396447, 0.0155353155]```

**PyTorch std**

```
[0.31951272, 0.06554735, 0.07942256, 0.18498649, 0.1079011 ,0.02935711, 0.02307329, 0.08221132, 0.03619578, 0.06809596, 0.08577741, 0.01684438]
```

Numerically it seems correct, so my current issue might be with ONNX export or elsewhere.

**Comment by kali** [comment]
Have you tried onnxruntime ?

**Comment by stillonearth** [comment]
Confirming, onnxruntime performance is similar to tract.

**Comment by kali** [comment]
Skimming the SAC documentation, I've spotted this:

```
Our SAC implementation uses a trick to improve exploration at the start of training. For a fixed number of steps at the beginning (set with the start_steps keyword argument), the agent takes actions which are sampled from a uniform random distribution over valid actions. After that, it returns to normal SAC exploration.
```

Is it possible that the exported network retains the "start_steps" variant instead of the real "exploitation" one we want ? Knowing how brittle and hackish pytorch export is, i would not be surprised.

**Comment by stillonearth** [comment]
It seems the problem with the invocation of model, because numerically
tract and onnxruntime converge.

пт, 30 дек. 2022 г., 19:25 Mathieu Poumeyrol ***@***.***>:

**Comment by kali** [comment]
... or the onnx export, I think.

**Comment by stillonearth** [comment]
In works fine. The problem was with model invocation method. According to `stable_baselines3.SAC` source code, model outputs are [scaled back](
https://github.com/DLR-RM/stable-baselines3/blob/4fa17dcf0f72455aa3d36308291d4b052b2544f7/stable_baselines3/common/policies.py#L263). Once scaling is implemented the model works exactly as PyTorch model. 

https://user-images.githubusercontent.com/97428129/210611507-efd8ac10-2b5c-4d6b-ad55-f0243d62266b.mp4
